### PR TITLE
feat(migration): #MZI-241 migration Vertx

### PR DESCRIPTION
**Pull request by Glama107**
## Describe your changes

Migration de Vertx 3 vers Vertx 4

## Issue ticket number and link

https://jira.support-ent.fr/browse/MZI-241

## Checklist before requesting a review (magic string, indentation, comment/documentation...)

- [x] I have detailed the tests to do in my feature/fix in order to prevent consequents regressions (must specify in **Checklist tests**)
- [x] I have performed a self-review of my code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation (API Doc etc...) - (must specify in **Description** for target version)
- [ ] If it is a consequent feature, I have added thorough tests.
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] Any dependent changes have been added to this project (must specify in **Description**)

### CODE REVIEW

File: zimbra/src/main/java/fr/openent/zimbra/filters/AccessibleDocFilter.java [hunk 1]
```diff
@@ -34,21 +35,26 @@ public class AccessibleDocFilter implements ResourcesProvider {
...
         String id = request.params().get("idAttachment");
 
         List<DBObject> groups = new ArrayList<>();
-        groups.add(QueryBuilder.start("userId").is(user.getUserId()).get());
-        for (String gpId : user.getGroupsIds()) {
-            groups.add(QueryBuilder.start("groupId").is(gpId).get());
+
+        List<Bson> groupFilters = new ArrayList<>();
+        groupFilters.add(Filters.eq("userId", user.getUserId()));
+        for (String groupId : user.getGroupsIds()) {
```
Comment (Alex-Delaveau) [inline zimbra/src/main/java/fr/openent/zimbra/filters/AccessibleDocFilter.java]: Utilisation de Stream plutot que la boucle : 
```
Bson subQuery = Filters.or(
    Filters.eq("userId", user.getUserId()),
    Filters.or(user.getGroupsIds().stream()
        .map(groupId -> Filters.eq("groupId", groupId))
        .toArray(Bson[]::new))
);
```

File: zimbra/src/main/java/fr/openent/zimbra/helper/PromiseHelper.java [hunk 2]
```diff
@@ -17,18 +17,22 @@ private PromiseHelper() {
         throw new IllegalStateException("Utility PromiseHelper class");
     }
 
-    public static <T> Handler<Either<String, T>> handlerJsonObject(Promise<T> promise) {
+    public static <T> Handler<Either<String, T>> handlerJsonObject(Promise<T> promise, String errorMessage) {
```
Comment (Alex-Delaveau) [inline zimbra/src/main/java/fr/openent/zimbra/helper/PromiseHelper.java:20]: errorMessage est il vraiment utile ? Utiliser un Optional?

File: zimbra/src/main/java/fr/openent/zimbra/service/data/SoapZimbraService.java [hunk 3]
```diff
@@ -185,105 +186,121 @@ private JsonObject prepareJsonRequest(JsonObject params) {
...
                 httpClient = HttpClientHelper.createHttpClient(vertx);
             }
-            String finalUrl = params.getBoolean(PARAM_ISADMIN) ? zimbraAdminUri : zimbraUri;
-            HttpClientRequest request;
-            Handler<HttpClientResponse> handlerRequest = zimbraRequestHandler(future.future());
-            request = httpClient.postAbs(finalUrl, handlerRequest);
-            request.setChunked(true);
-            if(params.getBoolean(PARAM_IS_AUTH, true) && params.containsKey(PARAM_AUTH_TOKEN)) {
-                request.putHeader(HEADER_COOKIE, COOKIE_AUTH_TOKEN + "=" + params.getString(PARAM_AUTH_TOKEN));
+            String finalUrl = Boolean.TRUE.equals(params.getBoolean(PARAM_ISADMIN)) ? zimbraAdminUri : zimbraUri;
```
Comment (Alex-Delaveau) [inline zimbra/src/main/java/fr/openent/zimbra/service/data/SoapZimbraService.java]: Cela ne sert à rien de tester si true == true : 
`Boolean.TRUE.equals(params.getBoolean(PARAM_ISADMIN))`

File: zimbra/src/main/java/fr/openent/zimbra/service/data/SoapZimbraService.java [hunk 4]
```diff
@@ -185,105 +186,121 @@ private JsonObject prepareJsonRequest(JsonObject params) {
...
+
+            RequestOptions requestOptions = new RequestOptions()
+                    .setAbsoluteURI(finalUrl)
+                    .setMethod(HttpMethod.POST)
+                    .setHeaders(new HeadersMultiMap());
+
+            if (Boolean.TRUE.equals(params.getBoolean(PARAM_IS_AUTH, true)) && params.containsKey(PARAM_AUTH_TOKEN)) {
```
Comment (Alex-Delaveau) [inline zimbra/src/main/java/fr/openent/zimbra/service/data/SoapZimbraService.java]: test true == true inutile

**Review by Alex-Delaveau — CHANGES_REQUESTED** [review]

**Review by Alex-Delaveau — APPROVED** [review]

File: zimbra/src/main/java/fr/openent/zimbra/Zimbra.java [hunk 5]
```diff
@@ -99,5 +100,8 @@ public void start() throws Exception {
         } catch (ParseException e) {
             log.warn("Mailer Cron deactivated");
         }
+
+        startPromise.tryComplete();
+        startPromise.tryFail("[Zimbra@Zimbra::start] Fail to start ApiZimbra");
```
Comment (qperie) [inline zimbra/src/main/java/fr/openent/zimbra/Zimbra.java]:
```suggestion
        startPromise.tryFail("[Zimbra@Zimbra::start] Fail to start Zimbra");
```

**Review by qperie — APPROVED** [review]

**Pull request merged**